### PR TITLE
Envisioned process for onboarding new tenants

### LINKED ISSUE #30: ADR covering how we anticpate onboarding a new tenant will work
This ADR documents how we anticipate the onboarding of a new tenant will work. This depends on a currently non-existent management plane API being in place, both for Marain.Instance and for individual Marain services that need to participate. This is not on the short-term roadmap for implementation; the short term approach will be documented in a separate ADR.

### PR DESCRIPTION
**Pull request by jongeorge1**
Resolves #30

### CODE REVIEW

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: `to some central thing`

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: Its!

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: I think it is important to make clear that the Client is *unaware* of (though may be able infer) the use of the FooBar service in two of those cases as they are an implementation detail of the services on which the client depends. You mention "is able to access those services" (or words to that effect) in a paragraph above and that is incorrect; "indirectly makes use of" is more accurate.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: See comment on line 260. I am concerned to make clear that the Client cannot *make use of* those services (directly or indirectly) but that those services are used on their behalf, and without their knowledge (even if they can infer it because of leaky implementation details such as not putting a facade over the operations API).

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: That is my grammatical Achilles' heel.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: I've added some additional wordage which hopefully reinforces that point.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: I've added some additional wordage which hopefully reinforces that point.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: It is not able to "indirectly use the service  via Workflow" 

It is more like 

`Workflow, in its implementation, uses Operations on Litware's behalf. That is something that Litware may be completely unaware of. In fact, Workflow may choose to dispense of its use of Operations in future, and, providing it maintained its current public API, that would be transparent to Litware. However, Litware requires the use of Operations in its own implementation. So, the process we went through for Workflow is repeated...
`

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (HowardvanRooijen) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: If would be useful if when you mentioned other ADRs they were hyperlinked for ease of reference.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (HowardvanRooijen) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: Is it just licensing? It appears that Licensing & Metering (for billing) are often bedfellows...

Comment (HowardvanRooijen) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: Also - what's your definition of licensing? Are you talking about access control? Rights management? Different feature tiers or SKUs? It feels like this might need a better definition.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (HowardvanRooijen) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: Isn't this the sort of scenario you want to use pub / sub / event grid for? Broadcast a "enroll" event and all services subscribe and enroll the tenant?

**Comment by mwadams** [comment]
If you read on, it is more a graph-orientated thing, rooted by the  direct-grant services into which you are enrolling right now. So definitely not "broadcast".

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (HowardvanRooijen) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: Again, If you broadcast a pub/sub message, any dependent service could also broadcast a message for its dependencies (and so on, and so forth).

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (HowardvanRooijen) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: I'm not sure that a direct service-to-service API call is the best strategy in a cloud hosted environment. It feels brittle and not very resilient.

**Comment by mwadams** [comment]
I agree with that point about direct service-to-service call. I think that some kind of accept-into-a-queue-and-process-it mechanism is more scalable at the expense of an extra moving part. But it is "queued request, wait for response" rather than broadcast. (And you'd use Marain.Operations to implement that :))

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: Yes - the problem is that the two related ADRs here are currently in separate PRs! Once one of them is merged back, I'll add the links as part of the other.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: The point with this is that if the dependency already exists, we are not introducing additional coupling.

So for example, with the Workflow service, which depends upon Operations, Workflow is already making direct service-to-service API calls to Operations to fulfil its own requests. Adding another touchpoint (the call to enrollment) doesn't decrease the resiliance.

Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: Also, if we were doing some kind of _accept-into-a-queue-and-process-it mechanism_ (as per @mwadams comment below), then we would normally hide this implementation detail from the caller - the enrollment endpoint would return a 202 and, as you say, use the operations service to manage the long running operation. We'd still end up with the direct service-to-service dependency, right?

Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: Or was the concern more along the lines of "what happens when part of this chain of enrollments fails?" Which is definitely an issue that we'll need to consider when we get to implementation.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: Yes - billing, metering, licensing (and authorization) are three (four) different things. It is perhaps better to call out that we are not really touching on those topics here. Merely "enrollment in the service". And we have worked hard to keep them orthogonal.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: Correct. I think Howard was describing the implementation mechanism rather than the concept, but the confusion came in the suggestion of "broadcast" which is specifically not what is happening.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (idg10) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: You seem to be referring to ADR 0006, but according to the filename, _this_ is ADR 0006.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (idg10) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: It's not obvious to me that to

> require all Marain services to have a control plane API 

is necessarily

> the simplest way for this to work 

Have we explored the possibility that for each service, we just write some sort of manifest describing the requirements? That we we only need one bit of code that knows how to configure a tenant, rather than N things (where N is the number of services). And that one bit of code could, to start with at least, be a console app rather than a thing that lives inside a Function, which would make it easier to develop and debug.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (idg10) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: However...can Litware be completely unaware of Workflow's use of Operations?

Practically speaking, the fact that Workflow hands out what are quite clearly URLs into the Operations service means that even if we don't tell Litware what's happening, they will most likely deduce that Operations is involved here. But there's a deeper issue here.

Hypothetically, yes Litware could be completely unaware of it in the same way that they could also be completely unaware that Workflow uses CosmosDB. That's also an implementation decision that Workflow could choose to dispense of in the future, much as it could dispense of its use of Operations. But if we are exposing to customers the fact that Workflow uses CosmosDB (and if we're contemplating a 'bring your own storage' model, then we definitely are) I don't see that it makes sense to do that and yet not expose the fact that this also means they're using Azure Blob Storage via the Storage service.

Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: I was expecting that we would expose that they are using blob storage in this scenario. In the workflow -> operations example, when enrolling into workflow I was expecting to have to gather config for both services - see line ~99:

_Firstly, we will request the list of required configuration to enroll a tenant for use with Workflow. Workflow "knows" it needs storage configuration, and because it's dependent on Operations and FooBar, it also requests the required configuration for those services. Operations does the same with FooBar, so the total list of required configuration for Workflow is the sum of those four things: Workflow storage config, Operations storage config, FooBar storage config when invoked from Workflow, and FooBar storage config when invoked from Operations._

If they are using Marain tenanted storage, they won't actually care and we will just use defaults for all of those things. If they are bringing their own, we will have to go through this with them as part of the onboarding/enrollment process, which will necessarily be more interactive due to the need to set up permissions, get connection strings, etc, etc.

So, whilst they _could_ use Workflow in blissful ignorance of the fact that it's using Operations , we aren't trying to hide what's going on for them and in some cases we might actually require them to know.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (HowardvanRooijen) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: It's not just failure cases, but also high load or unavailability scenarios. With provisioning / enrollment processes, it's about expressing the desired state, but also being able to deal with the asynchronicity of failure much further down the dependency chain.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: I'm not saying we are "hiding" it from them exactly, though the fact that we are giving them the opportunity to supply us with a Blob storage account that we want to use for storing long operation details doesn't necessarily indicate that we exposing the fact that we are using "Marain Operations" to do that. In fact, I probably *would* avoid making that clear.

Also, per @idg10 

> Practically speaking, the fact that Workflow hands out what are quite clearly URLs into the Operations service means that even if we don't tell Litware what's happening, they will most likely deduce that Operations is involved here.

There is no reason we should hand out URLs that are "quite clearly into the operations service"; providing an APIM facade over that would probably be better (so that we don't have the "on-behalf" tenant ID in an external URL, for example). That way we don't leak that implementation detail, and are protected against versioning issues.

However, I *do not* want to imply that the client in any way "uses" the Operations service. It can't. Workflow happens to return a long operation payload of the same schema as the operations service, and it currently uses `Marain.Operations` to do that. But that *is* an implementation detail.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (HowardvanRooijen) [inline docs/adr/0006-process-for-onboarding-new-tenants.md:294]: Are we easily able to determine the entire dependency tree, so that if we are decommissioning all of the tenants, we don't have orphans, because we've lost track of them?

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (HowardvanRooijen) [inline docs/adr/0006-process-for-onboarding-new-tenants.md:23]: How do we check if a tenant namespace is available and not already reserved? How do we deal with the situation that there is a conflict?

Comment (mwadams) [reply docs/adr/0006-process-for-onboarding-new-tenants.md:23]: The identity which is used is a GUID - the display name does not have to be unique.

Vanity URLs and rewriting provided over the guid for the tenant are obviously unique in the usual way.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [reply docs/adr/0006-process-for-onboarding-new-tenants.md:294]: Good question - we need to verify if DELETE of a child in its parent cascades that delete (which it should but I bet it doesn't) and when we get to this hypothetical offboarding thing, that it offboards all the children too.

You can walk that dependency tree, though.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: We've abandoned this now anyway for a somewhat different approach that only really involves the envisaged management API and the Tenancy service. We will still need to give careful thought to error handling, recovery, etc, but we no longer have to deal with an arbitrarily long chain of service-to-service dependencies.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: I think the wording now reflects that!

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: Various out-of-band discussions have taken place and I've updated the document to reflect this simpler approach. I have also refrained from using the phrase "the simplest way for this to work" :)

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: To make this --to-- work

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (mwadams) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: Fully hosted comes in two varieties. "Fully hosted in shared storage" and "fully hosted in isolated storage" which is like "bring your own storage" but we deploy the storage accounts for you. I think "fully hosted in isolated storage" will be more common than BYOS.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (idg10) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: > we can extend the process to provide support a way in which services can be notified of new enrollments 

I think the word "support" needs to go here? (You don't want both "provide" and "support" in there).

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (idg10) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: Do we need to be explicit about the fact that this includes indirect dependencies? (I think the example goes on to cover this, so maybe not.)

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (idg10) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: >There is no reason we should hand out URLs that are "quite clearly into the operations service"; providing an APIM facade over that would probably be better (so that we don't have the "on-behalf" tenant ID in an external URL, for example). That way we don't leak that implementation detail, and are protected against versioning issues.

I can think of one reason: hiding the tenant via the APIM layer adds a whole load of extra complexity in APIM. Imagine a world where I'm using 3 Marain services, each of which uses Operations as an implementation detail and hands out URLs to the client to operations in progress. There will now be 3 distinct on behalf of tenant ids. Somehow, the logic in APIM that conceals these needs to work out for each incoming request to get the status of one of these operations what the tenant id is. I don't know how it's going to do that if we stripped it out. I guess we could encrypt the URL path when rewriting the Location header of any response and decrypt it when the relevant GET requests come back in, but this seems like quite a lot of extra complexity to bake into the APIM layer to me.

More broadly, solutions of the form "We'll just handle that in APIM" may look architecturally satisfying, but I think that's only because they remove the problems from view. My concern is that the APIM layer will become a festering rat's nest.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: Made a slight tweak to the wording to make this clearer.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (idg10) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: Shouldn't this be "licensing"? (That's how the DVLA spell it, at any rate.)

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (idg10) [inline docs/adr/0006-process-for-onboarding-new-tenants.md]: Again, "licensing" I think.

**Review by idg10 — DISMISSED** [review]
Just the typo Matthew already identified and the two "licencing"s.

File: docs/adr/0006-process-for-onboarding-new-tenants.md
Comment (jongeorge1) [reply docs/adr/0006-process-for-onboarding-new-tenants.md]: Bah

**Review by mwadams — APPROVED** [review]

**Pull request merged**
